pangeo_forge_recipes/transforms.py

```diff
@@ -383,20 +383,86 @@ class StoreToZarr(beam.PTransform, ZarrWriterMixin):
+    target_chunk_size: Optional[Union[str, int]] = field(
+        default=None
+    )  # ? Should we provide a default?
+    size_tolerance: float = 0.2
+
+    def __post_init__(self):
+        # if none of the chunking parameters is specified, set the default behavior
+        # of target_chunks={}
+        if all(
+            a is None
+            for a in (self.target_chunks, self.target_chunk_size, self.target_chunks_aspect_ratio)
+        ):
+            self.target_chunks = {}
+
+        # check that not both static and dynamic chunking are specified
+        if self.target_chunks is not None and (
+            self.target_chunk_size is not None or self.target_chunks_aspect_ratio is not None
+        ):
+            raise ValueError(
+                (
+                    "Cannot specify both target_chunks and "
+                    "target_chunk_size or target_chunks_aspect_ratio."
+                )
+            )
+
+        # if dynamic chunking is specified, make sure both target_chunk_size
+        # and target_chunks_aspect_ratio are specified
+        if self.target_chunks is None and not (
+            self.target_chunk_size is not None and self.target_chunks_aspect_ratio is not None
+        ):
+            raise ValueError(
+                (
+                    "Must specify both target_chunk_size and "
+                    "target_chunks_aspect_ratio to enable dynamic chunking."
+                )
+            )
+
+    def determine_target_chunks(self, schema: XarraySchema) -> Dict[str, int]:
+        # if dynamic chunking is chosen, set the objects target_chunks here
+        # We need to explicitly cast the types here because we know
+        # that for the combinations chosen here (dynamic vs static chunking)
+        # the type of the input is never None (our __post_init__ method takes
+        # care of raising errors if this is the case).
+        if self.target_chunks_aspect_ratio is not None:
+            target_chunks = dynamic_target_chunks_from_schema(
+                schema,
+                target_chunk_size=cast(Union[int, str], self.target_chunk_size),
+                target_chunks_aspect_ratio=cast(Dict[str, int], self.target_chunks_aspect_ratio),
+                size_tolerance=self.size_tolerance,
+            )
+        else:
+            target_chunks = cast(Dict[str, int], self.target_chunks)
+        return target_chunks
 
     def expand(self, datasets: beam.PCollection) -> beam.PCollection:
         schema = datasets | DetermineSchema(combine_dims=self.combine_dims)
-        indexed_datasets = datasets | IndexItems(schema=schema)
-        rechunked_datasets = indexed_datasets | Rechunk(
-            target_chunks=self.target_chunks, schema=schema
+        target_chunks = beam.pvalue.AsSingelton(
+            schema
+            | beam.Map(
+                self.determine_target_chunks
+            )  # Would beam.pvalue.AsSingleton(self.determine_target_chunks(schema)) work?
```

Comment:
Or

```python
target_chunks = self.determine_target_chunks(beam.pvalue.AsSingleton(schema))
```

?

Reply:
Oh right, that makes a lot more sense. Changed.

Reply:
Oh wait getting [this](https://github.com/leap-stc/cmip6-leap-feedstock/actions/runs/5744818073/job/15571779448?pr=4#step:5:46) over at the CMIP6 recipes.

Reply:
I now settled on `target_chunks = beam.pvalue.AsSingleton(schema | beam.Map(self._get_target_chunks))` which seems to work, while just applying `AsSingelton` to the schema did not.